VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,537 @@
+
+- Go modules https://golang.org/ref/mod#versions use semver versions with
+  specific version resolution algorithms.
+
+- Haskell Package Versioning Policy https://pvp.haskell.org/ provides a notation
+  similar to this specification based on a modified semver with extra notations
+  such as star and caret.
+
+
+Note that there is a closely related problem: the way two versions are compared
+as equal, lesser or greater is complex and package types may have evolved their
+own peculiar version comparison procedure. Semver is a prominent specification
+in this domain but this is just one of the many versioning schemes. Semver,
+Debian, RPM, PyPI,  Rubygems, and Composer all have their own subtly different
+ways of comparing two versions even though they may seem highly similar at first
+glance.
+
+
+Solution
+---------
+
+A solution to the many version range syntaxes is to design a new notation to
+unify them all with:
+
+- a mostly universal and minimalist notation to express the version range
+  notations from many different package types and ecosystems.
+
+- the package type-specific definitions to normalize existing range expressions
+  to this common notation .
+
+- the designation of which algorithm or procedure to use when comparing two
+  versions such that it is possible to resolve if a version is within or
+  outside of a version range.
+
+We call this solution "version range specifier" or "vers" and it is described
+in this document.
+
+
+Version range specifier
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A version range specifier (aka. "vers") is a URI string using the ``vers``
+URI-scheme with this syntax::
+
+   vers:<versioning-scheme>/<version-constraint>,<version-constraint&version-constraint>,...
+
+For example to define a set of versions that contains either version ``1.2.3``,
+or any versions greater than or equal to ``2.0.0`` but less than ``5.0.0`` using
+the `semver` versioning scheme, the version range specifier will be::
+
+    vers:semver/1.2.3,>=2.0.0&<5.0.0`
+
+Each ``<version-constraint>`` in the comma-separated list is either a simple
+constraint such as::
+
+    <comparator:version>
+
+Or a composite constraint grouping multiple ``<version-constraint>`` joined by
+an ampersand such as:
+
+    <comparator:version>&<comparator:version>...
+
+The comma is a logical `OR` and the ampersand is a logical `AND`.
+
+A version range specifier is therefore an "OR of ANDs". This is also called a
+"disjunctive normal form" in boolean logic.
+See https://en.wikipedia.org/wiki/Disjunctive_normal_form for details.
+
+``vers`` is the URI-scheme and is an acronym for "VErsion Range Specifier". It
+has been selected because it is short, obviously about version and available
+for a future formal registration for this URI-scheme at the IANA registry.
+
+
+``<versioning-scheme>``
+------------------------
+
+The ``<versioning-scheme>`` (such as ``semver``,
+``debian``, etc.) determines:
+
```

Comment:
To my understanding, many of the version range schemas are not standards, they are policies or conventions. What happens when those conventions change over time? For example, if in a year, Debian changes how they represent relationships between packages, will that alter the meaning of every `vers:debian` being used?

Additionally, what happens when one of the schemes adopts another scheme? For example, if Maven were to adopt semver in the future?

Reply:
@stevespringett hey!
Thank you ++ for the feedback. That's a good point. 

Even though there may not always be a spec or clear documentation there is always code in the package management tools that provide something IMHO as good as a spec.

That said, to your point we cannot trust that upstream ecosystems will not change their ways. 

A few thoughts:
1. If they do this in a non-compatible way, this will wreak havoc on their ecosystem. Not a reassuring thought, but at least this makes these events less likely. But they can happen. See Go for instance.
2. When I helped cleanup the linux kernel licensing using SPDX identifiers there was a concern if the SPDX were to change. The solution was simply to incorporate the few bits we depended on in the kernel process documentation thus making the doc self standing (and this was a good decision as some key license identifiers were deprecated mid-way of this work). To make it short, we can incorporate these external specs as needed to "freeze" them in this spec.
3. And if they change in the future in a non-compatible, they would essentially create a new versioning scheme, and we could also create a new identifier for this versioning scheme such as `vers:debian2`, or `vers:maven-semver`

Unrelated, this makes me think that  `vers:semver` is not a versioning scheme IMHO ... I need to update the text: `node-semver` is rather it as used in a `package.json`: e.g. the version are `semver` but the version ranges are node-specific.

Reply:
> Unrelated, this makes me think that vers:semver is not a versioning scheme IMHO ... I need to update the text: node-semver is rather it as used in a package.json: e.g. the version are semver but the version ranges are node-specific.


I evolved this and now there is rather almost a 1-to-1 between a purl type and a vers scheme

Reply:
Just as another example of a convention that changed over time is the most widely used database in the world (probably billions of instances, would it be only on our phones, including the pre-smartphone area): **sqlite**.

[In 2015 it changed its versioning scheme to semver](https://www.sqlite.org/versionnumbers.html). And with a twist to make matter worst: the downloadable archives are not named sqlite-MAJOR.MINOR.PATCHLEVEL, but sqliteMAJORMINORPATCHLEVELWITHPADDINGZEROS. For example [3.37.0 is 3370000](https://www.sqlite.org/download.html). (It also uses its own CVS, Fossil, not git, but that's another concern.)